### PR TITLE
HTML Search: fixup: include partially-matched document titles in search results.

### LINKED ISSUE #12040: HTML Search: partially-matched titles are not included in search results.
### Describe the bug

As-written, the search code for Sphinx appears intended to give a [score of `7`](https://github.com/sphinx-doc/sphinx/blob/574519900ee7eb8b2a1422288139bcc20b56c6d9/sphinx/themes/basic/static/searchtools.js#L43) to [partial-matches](https://github.com/sphinx-doc/sphinx/blob/574519900ee7eb8b2a1422288139bcc20b56c6d9/sphinx/themes/basic/static/searchtools.js#L484-L487) of the user query against document titles.  This compares to a [score of `15`](https://github.com/sphinx-doc/sphinx/blob/574519900ee7eb8b2a1422288139bcc20b56c6d9/sphinx/themes/basic/static/searchtools.js#L42) when one of the terms from the user query [matches a title term exactly](https://github.com/sphinx-doc/sphinx/blob/574519900ee7eb8b2a1422288139bcc20b56c6d9/sphinx/themes/basic/static/searchtools.js#L475).

However: there is a bug.  The code uses the variable named `word` that relates to one of the words from the user's original query (`philos` in the repro case provided below) instead of the partially-matching title term found in the index (`philosphi` -- the `i` suffix is not a typo, that's due to [term stemming](https://en.wikipedia.org/wiki/Stemming)).

### How to Reproduce

Self-construct the Sphinx project's HTML documentation from commit 574519900ee7eb8b2a1422288139bcc20b56c6d9 and host this using a local webserver (for example, by running `python -m http.server -b 127.0.0.1` from the documentation build output directory).

Visit the locally-hosted project documentation using a web browser, and perform a search (keyboard shortcut: `/`) for the term `philos`.

Expected outcome: the [`Appendix: Deploying a Sphinx project online`](https://github.com/sphinx-doc/sphinx/blob/574519900ee7eb8b2a1422288139bcc20b56c6d9/doc/tutorial/deploying.rst) tutorial should be listed in the search results.

Actual outcome: the tutorial is not included in the search results.

### Environment Information

```text
Platform:              linux; (Linux-6.6.15-rt-amd64-x86_64-with-glibc2.37)
Python version:        3.11.8 (main, Feb  7 2024, 21:52:08) [GCC 13.2.0])
Python implementation: CPython
Sphinx version:        7.3.0+/574519900
Docutils version:      0.20.1
Jinja2 version:        3.1.3
Pygments version:      2.17.2
```

### Sphinx extensions

```python
extensions = ['sphinx.ext.autodoc', 'sphinx.ext.doctest', 'sphinx.ext.todo',
              'sphinx.ext.autosummary', 'sphinx.ext.extlinks',
              'sphinx.ext.intersphinx',
              'sphinx.ext.viewcode', 'sphinx.ext.inheritance_diagram',
              'sphinx.ext.coverage']
```

### Additional context

Discovered during code review of #11958.

### PR DESCRIPTION
**Pull request by jayaddison**
### Feature or Bugfix
- Bugfix

### Purpose
- Fixes a bug that prevents partially-matched titles from Sphinx documentation projects from appearing during search results (for example, a section titled `sphinx_utils module` _should_ be found when searching for `sphinx`, but currently is not).

### Detail
- The bug is that after finding a partially-matching title, a title-term-index lookup was performed using the `word` from the user query (`sphinx`, to continue the example) instead of the partially-matching term (`sphinx_utils`).

### Relates
- Depends upon **either**:
  - #11958
  - **OR**
  - #12037
- Resolves #12040.

### CODE REVIEW

File: sphinx/themes/basic/static/searchtools.js [hunk 1]
```diff
@@ -478,12 +478,12 @@ const Search = {
...
         Object.keys(terms).forEach((term) => {
-          if (term.match(escapedWord) && !terms[word])
+          if (term.match(escapedWord) && terms[word] === undefined)
             arr.push({ files: terms[term], score: Scorer.partialTerm });
         });
         Object.keys(titleTerms).forEach((term) => {
-          if (term.match(escapedWord) && !titleTerms[word])
-            arr.push({ files: titleTerms[word], score: Scorer.partialTitle });
+          if (term.match(escapedWord) && titleTerms[word] === undefined)
+            arr.push({ files: titleTerms[term], score: Scorer.partialTitle });
```
Comment (wlach) [inline sphinx/themes/basic/static/searchtools.js]: In case it's not immediately clear to others, this line is the actual fix.

**Review by wlach — APPROVED** [review]
Looks right to me

**Comment by jayaddison** [comment]
Thanks @wlach!

**Comment by picnixz** [comment]
Ok, I'm just picking a random PR on search to ask something:

@jayaddison @wlach Could you create an issue where you gather all the search-related PRs and the ones that should be merged (or alternatives)? like, which one should I focus first etc, whether one PR should be merged before another or not etc? They are a bit spread everywhere and it's awfully hard for me to pick the correct one.

**Comment by jayaddison** [comment]
Hi @picnixz - the [remaining pull requests](https://github.com/sphinx-doc/sphinx/pulls?q=is%3Aopen+is%3Apr+label%3A%22html+search%22+label%3A%22type%3Abug%22) can be considered independently of each other.

I've repurposed the `awaiting response` issue label to tag the pull requests that are awaiting review.

Logically I think #11578 would be the first one to merge, followed by #12041.

Two of the pull requests are competing/alternate solutions to the same problem; we think we've decided on our preferred pull request -- but even so, if possible, I'd like if someone with long-term experience of the search code to catch any mistakes with that (because my sense is that it could be trickier than some other changes to rollback/revert problems).

**Comment by picnixz** [comment]
> I've repurposed the awaiting response issue label to tag the pull requests that are awaiting review.

I can create another label for that (awaiting:review)

**Comment by picnixz** [comment]
So 12041 is actually this one. Just wondering, but the partially matched documents are put after those with exact match, right?

**Comment by picnixz** [comment]
Ok, then this PR is fine (just resolve the conflicts and I'll merge it)

**Comment by jayaddison** [comment]
Nearly-ready, but I think I'll block/wait on this until the JS search tests are refactored.

**Comment by picnixz** [comment]
@jayaddison I know you said that you would be offline but since nothing changed since last time (except conflicts and CHANGES), I assume that everything is fine, right? if so, I'll merge it this afternoon.

**Review by picnixz — APPROVED** [review]

**Comment by jayaddison** [comment]
@picnixz yep; this is ready and can be merged without other dependencies, so I felt it could be good to mark as ready for review while I'm taking a rest.  Thank you.

**Comment by picnixz** [comment]
Thank you both!

**Pull request merged**